File: .tmp_py4013/batch_18_dump/3970_shortest_path_with_at_most_k_consecutive_identical_characters.js
// LeetCode 3970 - Shortest Path With At Most K Consecutive Identical Characters
// https://leetcode.com/problems/shortest-path-with-at-most-k-consecutive-identical-characters/

var shortestPath = function(n, edges, labels, k) {
    const graph = Array.from({length: n}, () => []);
    for (const edge of edges) graph[edge[0]].push([edge[1], edge[2]]);
    const infinity = Number.MAX_SAFE_INTEGER / 4;
    const distances = Array.from({length: n}, () => new Array(k + 1).fill(infinity));
    distances[0][1] = 0;
    const pq = [[0, 0, 1]]; // [distance, node, run]
    const cmp = (a, b) => a[0] - b[0];
    while (pq.length) {
        pq.sort(cmp);
        const cur = pq.shift();
        const distance = cur[0], node = cur[1], run = cur[2];
        if (distance !== distances[node][run]) continue;
        if (node === n - 1) return distance;
        for (const e of graph[node]) {
            const to = e[0], weight = e[1];
            let nextRun = 1;
            if (labels[node] === labels[to]) nextRun = run + 1;
            if (nextRun > k) continue;
            const nextDistance = distance + weight;
            if (nextDistance < distances[to][nextRun]) {
                distances[to][nextRun] = nextDistance;
                pq.push([nextDistance, to, nextRun]);
            }
        }
    }
    return -1;
};

File: .tmp_rs4013/rust_check22/3970_shortest_path_with_at_most_k_consecutive_identical_characters.rs
struct Solution;
// LeetCode 3970 - Shortest Path With At Most K Consecutive Identical Characters
// https://leetcode.com/problems/shortest-path-with-at-most-k-consecutive-identical-characters/

use std::cmp::Reverse;
use std::collections::BinaryHeap;

impl Solution {
    pub fn shortest_path(n: i32, edges: Vec<Vec<i32>>, labels: String, k: i32) -> i64 {
        let n = n as usize;
        let k = k as usize;
        let mut graph = vec![Vec::new(); n];
        for edge in &edges {
            graph[edge[0] as usize].push((edge[1] as usize, edge[2]));
        }
        let labels = labels.into_bytes();
        const INFINITY: i64 = ((u64::MAX) >> 2) as i64;
        let mut distances = vec![vec![INFINITY; k + 1]; n];
        distances[0][1] = 0;
        let mut queue: BinaryHeap<Reverse<(i64, usize, usize)>> = BinaryHeap::new();
        queue.push(Reverse((0, 0, 1)));
        while let Some(Reverse((distance, node, run))) = queue.pop() {
            if distance != distances[node][run] {
                continue;
            }
            if node == n - 1 {
                return distance;
            }
            for &(to, weight) in &graph[node] {
                let next_run = if labels[node] == labels[to] { run + 1 } else { 1 };
                if next_run > k {
                    continue;
                }
                let next_distance = distance + weight as i64;
                if next_distance < distances[to][next_run] {
                    distances[to][next_run] = next_distance;
                    queue.push(Reverse((next_distance, to, next_run)));
                }
            }
        }
        -1
    }
}

fn main() {}
